pkg/core/blockchain.go

```diff
@@ -401,71 +401,45 @@ func (bc *Blockchain) persistBlock(block *Block) error {
 	if err := assets.commit(batch); err != nil {
 		return err
 	}
-	if err := bc.PutBatch(batch); err != nil {
+	if err := bc.memStore.PutBatch(batch); err != nil {
 		return err
 	}
 
-	bc.persistedHeight = block.Index
+	atomic.StoreUint32(&bc.blockHeight, block.Index)
 	return nil
 }
 
-// persist flushed current block cache to the persistent storage.
-func (bc *Blockchain) persist(ctx context.Context) (err error) {
+// persist flushed current in-memory store to the persistent storage.
+func (bc *Blockchain) persist(ctx context.Context) error {
 	var (
 		start     = time.Now()
 		persisted = 0
-		lenCache  = bc.blockCache.Len()
+		err       error
 	)
 
-	if lenCache == 0 {
-		return nil
-	}
-
-	bc.headersOp <- func(headerList *HeaderHashList) {
-		for i := 0; i < lenCache; i++ {
-			if uint32(headerList.Len()) <= bc.BlockHeight() {
-				return
-			}
-			hash := headerList.Get(int(bc.persistedHeight + 1))
-			if block, ok := bc.blockCache.GetBlock(hash); ok {
-				if err = bc.persistBlock(block); err != nil {
-					return
-				}
-				bc.blockCache.Delete(hash)
-				persisted++
-			} else {
-				// no next block in the cache, no reason to continue looping
-				break
-			}
-		}
+	persisted, err = bc.memStore.Persist(bc.Store)
+	if err != nil {
+		return err
 	}
-
-	select {
-	case <-ctx.Done():
```

Comment:
wow. Could you please comment this? I mean why we don't need to write to ctx anymore?

Reply:
We weren't writing there, we were waiting for `headersOp` to finish. And while we were waiting, there was a chance for `ctx.Done()` to come first and if it really comes first then there is no point in waiting for `bc.headersOpDone`, that's why there was a `select` there. Now we just don't do `headersOp`, don't wait for `bc.headersOpDone` so there is no point in checking for `ctx.Done()`.

Reply:
that makes sense! ty!